src/librustc_mir/transform/check_unsafety.rs

```diff
@@ -337,22 +339,38 @@ impl<'a, 'tcx> UnsafetyChecker<'a, 'tcx> {
                     match violation.kind {
                         UnsafetyViolationKind::GeneralAndConstFn
                         | UnsafetyViolationKind::General => {}
-                        UnsafetyViolationKind::BorrowPacked(_) => {
+                        UnsafetyViolationKind::BorrowPacked => {
                             if self.min_const_fn {
                                 // const fns don't need to be backwards compatible and can
                                 // emit these violations as a hard error instead of a backwards
                                 // compat lint
                                 violation.kind = UnsafetyViolationKind::General;
                             }
                         }
+                        UnsafetyViolationKind::UnsafeFn => {
+                            bug!("`UnsafetyViolationKind::UnsafeFn` in an `Safe` context")
+                        }
                     }
                     if !self.violations.contains(&violation) {
                         self.violations.push(violation)
                     }
                 }
                 false
             }
-            // `unsafe` function bodies allow unsafe without additional unsafe blocks
+            // With the RFC 2585, no longer allow `unsafe` operations in `unsafe fn`s
+            Safety::FnUnsafe if self.tcx.features().unsafe_block_in_unsafe_fn => {
+                for violation in violations {
+                    let mut violation = *violation;
+
+                    // FIXME(LeSeulArtichaut): what to do with `UnsafetyViolationKind::BorrowPacked`?
```

Comment:
This seems like the major "to do" here.

I have to say I'm a bit surprised by all the moving and shuffling of things from vector to vector. I guess I need to look over how the code works a bit to see why it's setup this way (no doubt there's a good reason). 

That said, let's see, I guess we have to ask ourselves what behavior we *expect* here. My expectation is that `&foo.bar` where `bar` is a packed field is an unsafe operation and should behave like any other unsafe operation, *except* that -- when it is reported -- it will trigger as a lint (per the level of that lint).

I think what this implies to me is that perhaps it is wrong to treat an error within an unsafe function as as "unsafety violation kind", or wrong to treat `BorrowPacked` as a kind. i.e., they are basically orthogonal things. 

One could also have `UnsafeFnBorrowPacked` I suppose.

Reply:
Here is what I proposed earlier in RalfJung's review:
> I can see two possible solutions:
> - we can use the level defined for `unsafe_op_in_unsafe_fn`, but don't emit anything if `safe_packed_borrow` is `allow`ed
> - or we can use the lowest level of both lints

RalfJung noted that the latter seemed to make the most sense, and I think I agree. Does it seem OK to you?

Reply:
I don't think it does. As I noted above, I would expect that the two settings are kind of "orthogonal". In particular:

* if unsafe operations are permitted in unsafe fns (the new lint is allow), then I would expect that we never report a use of an unsafe field
* if unsafe operations are not permitted in unsafe fns (the new lint is warn), but if the lint for accessing packed fields is set to deny, then I would expect us to report with deny

So it's not quite the "minimum of the two".